crates/compiler/parse/src/src64.rs

```diff
@@ -0,0 +1,550 @@
+                //
+                // Note that on UNIX, bytes_read might be -1 if this was a file read error. This
+                // condition will catch that too, since we know file_size won't be (-1isize as usize)
+                // beacuse if it was, then this match would have taken the None branch due to
+                // (-1isize as usize) exceeding our maximum file size.
+                if bytes_read != file_size {
+                    return Err(FileErr::ReadErr);
+                }
+
+                // Safety: bytes_ptr came from an allocation of `capacity` bytes, it's had
+                // newlines filled at the end, and `file_size` bytes written over the rest.
+                let bytes =
+                    unsafe { core::slice::from_raw_parts_mut(buf.as_ptr() as *mut u8, capacity) };
+
+                Ok(Self { bytes })
+            }
+            None => Err(FileErr::FileWasTooBig(file_size)),
+        }
+    }
+}
+
+fn round_up_to_nearest_64(num: usize) -> usize {
+    // Round up to the nearest 64. (Writing this out as 64 - 1 so it's clearer where the numbers came from.)
+    // We can do saturating addition here rather than overflow checking, because if we overflow usize::MAX,
+    // we will most definitely be over the max source file size and return None anyway.
+    (num.saturating_add(64 - 1)) & !(64 - 1)
+}
+
+/// Safety: capacity must be a multiple of 64, and must be at least 64.
+unsafe fn allocate_and_pad_with_newlines(arena: &Bump, capacity: usize) -> Option<NonNull<u8>> {
+    // Compare capacity here instead of size because this file limit is based on what we can record row and line
+    // numbers for, and those can theoretically oveflow on the trailing newlines we may have added.
+    // This distinction will most likely come up in practice zero times ever, but it could come up in fuzzing.
+    if capacity > MAX_ROC_SOURCE_FILE_SIZE {
+        return None;
+    }
+
+    debug_assert!(capacity >= 64);
+    debug_assert!(capacity % 64 == 0);
+
+    // Safety: the rules we follow are https://doc.rust-lang.org/core/alloc/struct.Layout.html#method.from_size_align_unchecked
+    // `align` is valid because it's hardcoded, and we already rounded `capacity` up to something even bigger.
+    // We align it to 64B so that it's on cache line boundaries on many CPUs, which makes prefetching simpler.
+    let layout = unsafe { Layout::from_size_align_unchecked(capacity, Src64::BYTES_ALIGNMENT) };
+
+    // We have to use alloc_layout here because we have stricter alignment requirements than normal slices.
+    let buf_ptr: NonNull<u8> = arena.alloc_layout(layout);
+
+    // Branchlessly prefetch the first three 64-byte chunks, and the last chunk. This prevents a double cache miss:
+    // first a cache miss when we write the newlines to the end, and then a second cache miss when we start
+    // working on the beginning of the allocation. We can do further prefetches in the actual tokenization loop.
```

Comment:
Prefetching here seems like it would make a difference (for the reason mentioned in the comment), but once I have everything working I can try removing this and see whether it actually made a difference in practice.